northstar/src/npk/npk.rs

```diff
@@ -748,6 +758,107 @@ fn sign_hashes(key_pair: &Keypair, hashes_yaml: &str) -> String {
+fn set_selinux_contexts(root: &Path, context_type: &NonNullString) -> Result<(), Error> {
+    for dir_entry in WalkDir::new(root)
+        .follow_links(false)
+        .into_iter()
+        .filter_map(Result::ok)
+    {
+        let path = dir_entry.path();
+        let path_c = CString::new(path.as_os_str().as_bytes()).map_err(|_| {
+            Error::Selinux(format!(
+                "Failed to convert path to C string: {}",
+                &path.display()
+            ))
+        })?;
+
+        // Read existing selinux context
+        let context = read_selinux_context(&path_c).ok_or_else(|| {
+            Error::Selinux(format!(
+                "Failed to read selinux context from {}",
+                &path.display()
+            ))
+        })?;
+
+        // Replace selinux type of current file
+        let context = replace_selinux_context_type(&context, context_type).ok_or_else(|| {
+            Error::Selinux(format!(
+                "Failed to update selinux context of {}",
+                &path.display()
+            ))
+        })?;
+
+        // Write updated context
+        if !write_selinux_context(&path_c, &context) {
+            return Err(Error::Selinux(format!(
+                "Failed to write selinux context to {}",
+                &path.display()
+            )));
+        }
+    }
+    Ok(())
+}
+
+/// Read the selinux context from a path string
+fn read_selinux_context(path_c: &CStr) -> Option<String> {
+    let name_c = CString::new("security.selinux").unwrap();
+    let size = unsafe {
+        nix::libc::getxattr(
+            path_c.as_ptr() as *const nix::libc::c_char,
+            name_c.as_ptr() as *const nix::libc::c_char,
+            std::ptr::null_mut() as *mut core::ffi::c_void,
+            0,
+        )
+    };
+    if size == -1 {
+        return None;
+    }
+    let mut context_buf = vec![0u8; size.try_into().ok()?];
+    let size = unsafe {
+        nix::libc::getxattr(
+            path_c.as_ptr() as *const nix::libc::c_char,
+            name_c.as_ptr() as *const nix::libc::c_char,
+            context_buf.as_mut_ptr() as *mut core::ffi::c_void,
+            context_buf.len(),
+        )
+    };
+    if size == -1 {
+        return None;
+    }
+    let context = CStr::from_bytes_with_nul(&context_buf[..size.try_into().ok()?])
+        .ok()?
+        .to_str()
+        .ok()?
+        .to_string();
+    Some(context)
+}
+
+/// Replace the tag of a selinux context and return the updated version
+fn replace_selinux_context_type(context: &str, context_type: &NonNullString) -> Option<CString> {
+    let mut components: Vec<_> = context.split(':').collect();
+    if components.len() >= 3 {
+        components[2] = context_type;
+        CString::new(components.join(":").as_bytes()).ok()
+    } else {
+        None
+    }
+}
+
+/// Write selinux context to a path
+fn write_selinux_context(path_c: &CString, context: &CString) -> bool {
```

Comment:
Why is a `CString` used here? I would have expected a `&std::path::Path` or at least a `&CStr`.
`&CString` doesn't make sense.